### DISCUSSION TITLE
AttributeError: 'PosixPath' object has no attribute 'expanduser'

### DISCUSSION THREAD

**Discussion by gauravcanon** [pydantic/pydantic#4854] [Question]
I am trying to deploy Python FastApi on the Azure web app. Deployment is working as expected.

But after deployment Azure web app is showing error. Error is only appearing in Azure Webapp

```python
/home/LogFiles/2022_12_17_lw1sdlwk0006KE_default_docker.log  (https://example.scm.azurewebsites.net/api/vfs/LogFiles/2022_12_17_lw1sdlwk0006KE_default_docker.log)
2022-12-17T18:09:53.379366634Z     from app.core.log_file_handler import logs as Log
2022-12-17T18:09:53.379535236Z   File "/tmp/8dae0581e110892/app/core/log_file_handler.py", line 5, in <module>
2022-12-17T18:09:53.379544536Z     from app.core.utils import settings
2022-12-17T18:09:53.379548236Z   File "/tmp/8dae0581e110892/app/core/utils.py", line 154, in <module>
2022-12-17T18:09:53.379551936Z     settings = Settings()
2022-12-17T18:09:53.388338451Z   File "pydantic/env_settings.py", line 39, in pydantic.env_settings.BaseSettings.__init__
2022-12-17T18:09:53.388357051Z   File "pydantic/env_settings.py", line 73, in pydantic.env_settings.BaseSettings._build_values
2022-12-17T18:09:53.388362051Z   File "pydantic/env_settings.py", line 166, in pydantic.env_settings.EnvSettingsSource.__call__
2022-12-17T18:09:53.388365951Z AttributeError: 'PosixPath' object has no attribute 'expanduser'
```
Running fine locally on windows , linux , macos , only problem is happening in Azure web app This is my utils.py file

```python

load_dotenv()


class Settings(BaseSettings):
    """
    Get credentials values from .env file
    """

    # Environment
    environment: str = os.environ.get("ENVIRONMENT")

    # Database credentials
    db_username: str = os.environ.get("DB_USERNAME")
    db_password: str = os.environ.get("DB_PASSWORD")
    db_url: str = os.environ.get("DB_URL")
    db_name: str = os.environ.get("DB_NAME")

    title: str = "GTCRM-API"
    description: str = """ API 🚀 """
    version: str = "1.3.0"
    contact = {
        "name": "Appname",
        "url": "https://example.com/",
        "email": "support@example.com",
    }
    openapi_tags = tags_metadata
    docs_url: str = "/docs"
    redoc_url: str = "/redocs"
    openapi_url: str = "/openapi.json"
    disable_docs: bool = os.environ.get("ENVIRONMENT") != "development"

    @property
    def fastapi_kwargs(self) -> Dict[str, Any]:
        """This returns a dictionary of the most commonly used keyword
        arguments when initializing a FastAPI instance.

        If `self.disable_docs` is True, the various docs-related arguments
        are disabled, preventing spec from being published.
        """
        fastapi_kwargs: Dict[str, Any] = {
            "title": self.title,
            "description": self.description,
            "version": self.version,
            "contact": self.contact,
            "openapi_tags": self.openapi_tags,
            "docs_url": self.docs_url,
            "redoc_url": self.redoc_url,
            "openapi_url": self.openapi_url,
        }
        if self.disable_docs:
            fastapi_kwargs.update(
                {"docs_url": None, "openapi_url": None, "redoc_url": None}
            )
        return fastapi_kwargs

    class Config:
        env_file = ".env"


# create instance of class
settings = Settings()

s3_client = boto3.client(
    "s3",
    # some env declatrion is deleted above to keep code short below where listed
    aws_access_key_id=settings.aws_access_key_id,
    aws_secret_access_key=settings.aws_secret_access_key,
    region_name=settings.region_name,
)
```
Not able to understand what an error is . Any help will be great 
pydantic==1.9.1

Showing error on 
`env_path = Path(env_file).expanduser()`

```python
    def _read_env_files(self, case_sensitive: bool) -> Dict[str, Optional[str]]:
        env_files = self.env_file
        if env_files is None:
            return {}

        if isinstance(env_files, (str, os.PathLike)):
            env_files = [env_files]

        dotenv_vars = {}
        for env_file in env_files:
            env_path = Path(env_file).expanduser()
            if env_path.is_file():
                dotenv_vars.update(
                    read_env_file(env_path, encoding=self.env_file_encoding, case_sensitive=case_sensitive)
                )

        return dotenv_vars
        ```

**Comment by Viicos** [comment]
The purpose of the `BaseSettings` class is to use environment variables as a source for your variables, see the [docs](https://docs.pydantic.dev/usage/settings/) for more details. So you don't need to explicitly read them from `os.environ`. With the `Config` class you provided, your `Settings` class will read from both your .env file as well as environment variables. Note from the docs:

> Even when using a dotenv file, pydantic will still read environment variables as well as the dotenv file, **environment variables will always take priority over values loaded from a dotenv file.**


Regarding the `AttributeError: 'PosixPath' object has no attribute 'expanduser'`, you are probably running an outdated version of Python (<3.5), which seems unlikely as `pydantic==1.9.1` requires Python >=3.6.1. Could you tell us which version of Python you are running?

**Reply by gauravcanon** [reply]
@Viicos Thanks for looking into it Python version is 3.9.7

**Reply by Viicos** [reply]
Weird, the only place where I could see this is [here](https://forum.omz-software.com/topic/6243/missing-pathlib-expanduser/6). What about doing `import os, pathlib; print(os.path.dirname(pathlib.__file__))`? I suspect your deployment environment might be part of the reason

**Reply by okechukwu0127** [reply]
> @Viicos Thanks for looking into it Python version is 3.9.7

I found a way around this issue by doing some code refactoring


**Repalcing the following code**
```
ROOTDIR=${input:=$HOME}/invokeai
from pathlib import Path; 
print(Path('${ROOTDIR}').expanduser().resolve())
```


**with the following code**
```
ROOTDIR=${input:=$HOME}/invokeai
import  os; 
print(os.path.expanduser('${ROOTDIR}'));
```


I hope this work around works for you as it did for me

**Reply by dirobayisameka** [reply]
import os.path
path='C:\\Users\\hp\\Desktop\\my_photos'
data_dir=os.path.expanducer('C:\\Users\\hp\\Desktop\\my_photos')
it says AttributeError: module 'posixpath' has no attribute 'expanducer'

**Reply by dmontagu** [reply]
@dirobayisameka are you also trying to deploy using the Azure App Service? Since that was the only environment where the original poster had the issue, I'm wondering if maybe Azure did something funny here for "security purposes" and removed that attribute from the python they run. (I really know nothing about their web app running service, so maybe that's obviously not the issue.)

Something weird to me is that you are using what are clearly windows paths, yet it is attempting to access `expanduser` from `posixpath`, which indicates a posix system; in a windows install I think it should be importing `ntpath`. But maybe this is a WSL thing or similar?

Alternatively, are you using any other (non-standard) way of installing python?

**Comment by samuelcolvin** [comment]
This seems very weird, you must have a broken python installation, I'm afraid there's really nothing we can do about this on pydantic.

If more users get this, I guess we could catch a `AttributeError` on `expanduser()`, but no one else has reported  this.

The docs are pretty clear that the only error you should get from `expanduser()` is a `RuntimeError`.

**Reply by okechukwu0127** [reply]
I just got this same error now 

AttributeError: 'PosixPath' object has no attribute 'expanduser' 

even after running **python -m pip install pathlib** 

My **Python version is 3.10.5**

**Reply by Viicos** [reply]
@okechukwu0127 `pathlib` is not a pip package, it is part of the standard library. Could you please reproduce what I've said in this comment https://github.com/pydantic/pydantic/discussions/4854#discussioncomment-4475124?

**Reply by okechukwu0127** [reply]
> @okechukwu0127 `pathlib` is not a pip package, it is part of the standard library. Could you please reproduce what I've said in this comment [#4854 (reply in thread)](https://github.com/pydantic/pydantic/discussions/4854#discussioncomment-4475124)?

I was actually able to resolve my issue using a different aproach. I introduced  **os.path.expanduser()** and it worked for me

Kind regards

**Reply by Viicos** [reply]
Ok nice, just out of curiosity, are you running your program in a special environment? As it seems you're not the only one having this issue, there might be an underlying problem

**Comment by dirobayisameka** [comment]
ok thanks
